Add docs for servodriver. (#59)
servo/book: src/architecture/servodriver.md

We expose [two web-accessible methods](https://github.com/servo/servo/blob/3421185737deefe27e51e104708b02d9b3d4f4f3/components/script/dom/window.rs#L1189-L1203) for communicating async results to the webdriver server, `Window.webdriverCallback` and `Window.webdriverTimeout`.

### Input handler

When the webdriver server receives input action commands (eg. pointer or mouse inputs), it creates an action sequence and [dispatches them progressively](https://github.com/servo/servo/blob/3421185737deefe27e51e104708b02d9b3d4f4f3/components/webdriver_server/actions.rs#L110) to the compositor ([via the constellation](https://github.com/servo/servo/blob/3421185737deefe27e51e104708b02d9b3d4f4f3/components/constellation/constellation.rs#L4548-L4559)).
The compositor then [handles these events](https://github.com/servo/servo/blob/3421185737deefe27e51e104708b02d9b3d4f4f3/components/compositing/compositor.rs#L587-L601) the same as any input events received from the embedder (modulo [known bugs](https://github.com/servo/servo/issues/35394)).

## The test page scripts

The testdriver harness works by incorporating all of these elements together.

> The executor [opens a new window](https://github.com/servo/servo/blob/3421185737deefe27e51e104708b02d9b3d4f4f3/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py#L850)  and invokes an async script that [sets up a testdriver callback](https://github.com/servo/servo/blob/3421185737deefe27e51e104708b02d9b3d4f4f3/tests/wpt/tests/tools/wptrunner/wptrunner/executors/testharness_webdriver_resume.js) (which is `Window.webdriverCallback`).

Reviewer: ```suggestion
The executor [opens a new window](https://github.com/servo/servo/blob/3421185737deefe27e51e104708b02d9b3d4f4f3/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py#L850) and invokes an async script that [sets the testdriver callback](https://github.com/servo/servo/blob/3421185737deefe27e51e104708b02d9b3d4f4f3/tests/wpt/tests/tools/wptrunner/wptrunner/executors/testharness_webdriver_resume.js) to `Window.webdriverCallback` (`arguments[arguments.length - 1]`).
```

> This window creates [a message queue](https://github.com/servo/servo/blob/3421185737deefe27e51e104708b02d9b3d4f4f3/tests/wpt/tests/tools/wptrunner/wptrunner/executors/message-queue.js#L30) which is [read by the executor harness](https://github.com/servo/servo/blob/3421185737deefe27e51e104708b02d9b3d4f4f3/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py#L904-L905).

Reviewer: ```suggestion
This window creates [a message queue](https://github.com/servo/servo/blob/3421185737deefe27e51e104708b02d9b3d4f4f3/tests/wpt/tests/tools/wptrunner/wptrunner/executors/message-queue.js#L30), which is [read by the executor harness](https://github.com/servo/servo/blob/3421185737deefe27e51e104708b02d9b3d4f4f3/tests/wpt/tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py#L904-L905).
```

> Subsequently, each test that runs is opened as a new window—this allows them to use `opener.postMessage` to communicate with the original window, which allows the testdriver APIs to send messages that [represent webdriver API calls](https://github.com/servo/servo/blob/3421185737deefe27e51e104708b02d9b3d4f4f3/tests/wpt/tests/tools/wptrunner/wptrunner/testdriver-extra.js#L286-L296).

Reviewer: ```suggestion
Subsequently, each test that runs is opened as a new window; these windows can then use `opener.postMessage` to communicate with the original window, which allows the testdriver APIs to send messages that [represent WebDriver API calls](https://github.com/servo/servo/blob/3421185737deefe27e51e104708b02d9b3d4f4f3/tests/wpt/tests/tools/wptrunner/wptrunner/testdriver-extra.js#L286-L296).
```

## Debugging hints

Always start with the following RUST_LOG:

```
RUST_LOG=webdriver_server,webdriver,script::webdriver_handlers,constellation
```

> This usually makes clear if the expected API calls are being processed.

Reviewer: ```suggestion
This usually makes it clear if the expected API calls are being processed.
```

> When debugging input-related problems, add JS debugging lines that log the `getBoundingClientRect()` of whatever element is being clicked, then compare the coordinates received by the compositor when clicking manually vs. the coordinates received from the webdriver server.

Reviewer: ```suggestion
When debugging input-related problems, add JS debugging lines that log the `getBoundingClientRect()` of whatever element is being clicked, then compare the coordinates received by the compositor when clicking manually vs the coordinates received from the WebDriver server.
```
